platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/ZipUtils.java

```diff
@@ -0,0 +1,284 @@
+import static org.apache.commons.io.FileUtils.writeStringToFile;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Charsets;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.KeyGenerator;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.io.IOUtils;
+import org.codice.ddf.migration.MigrationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ZipUtils {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ZipUtils.class);
+
+  private static final String IMPORT_KEY_PADDING_ERROR =
+      "Import error: invalid padding used for import key {}.";
+
+  private static final String IMPORT_KEY_ALGORITHM_ERROR =
+      "Import error: invalid algorithm used for import key {}.";
+
+  private static final String EXPORT_KEY_PADDING_ERROR =
+      "Export error: invalid padding used for export key";
+
+  private static final String EXPORT_KEY_ALGORITHM_ERROR =
+      "Export error: invalid algorithm used for export key";
+
+  @VisibleForTesting static final String KEY_ALGORITHM = "AES";
+
+  @VisibleForTesting static final String CIPHER_ALGORITHM = "AES/ECB/PKCS5Padding";
+
+  @VisibleForTesting static final String DECRYPTION_KEY_FILE_NAME = "export.key";
+
+  @VisibleForTesting static final String CHECKSUM_FILE_EXTENSION = ".sha256";
+
+  @VisibleForTesting static final String CHECKSUM_DIGEST_ALGORITHM = "SHA-256";
+
+  /**
+   * Streams the contents of a {@link ZipInputStream} as a {@link Map} where the content of each
+   * {@link ZipEntry} is stored inside a key corresponding to each {@link ZipEntry}'s name
+   *
+   * @param in the {@link ZipInputStream} to retrieve the entries from
+   * @return the stream of entries
+   */
+  public static Map<ZipEntry, byte[]> entriesFrom(ZipInputStream in) {
+    ZipEntry entry;
+    Map<ZipEntry, byte[]> entries = new HashMap<>();
+    try {
+      while ((entry = in.getNextEntry()) != null) {
+        if (!entry.isDirectory()) {
+          byte[] content = IOUtils.toByteArray(in);
+          entries.put(entry, content);
+        }
+      }
+    } catch (IOException e) {
```

Comment:
@paouelle at the moment I'm kinda torn on this one. Part of me feels that it might be better to handle the IOExceptions from reading the next ZipEntry up front. I can also see the benefit of deferring it though. Im curious would you be alright if the deferral of the error handling could be implemented in a follow-on PR so that we can get the first iteration of the encrypted migration into the baseline first? 
If you feel this behavior is essential for this pr however I will work on finishing my implementation of a ZipFile-like class that will wrap the CipherInputStream for each entry instead of creating a ticket to iterate further on this approach

Reply:
My issue is that it changes the current design.

Reply:
I agree with @paouelle here. Even though we technically comply with the API's methods, we're changing the intended behavior, which is to abstract how the entries are stored and fail only when they are accessed, not at the very beginning.

If we were to replace the zip with a directory structure for instance, I would not expect the migration to fail at the very beginning if one of the files is missing or corrupted. Only when a migratable tries to access such a file. And even then, it should be up to the migratable to decide whether it should fail or not.

Switching to this implementation changes that behavior, which would have to change again if we were to postpone the change Patrick is asking for.

As a side-note, we may want to document this behavior somewhere in the API to make this clear and ensure that implementations always work consistently.

Reply:
@lessarderic that is a good point about how the behavior would look if the migration was changed to use a directory structure. I really hadn't thought of it that way, I was just thinking of the zip as a single entity. I will continue working on an implementation that will allow for this to work as intended.